### ISSUE TITLE
pinalikefruit - The owner/users must pay more than baseFee when creating a pool

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-09-Gitcoin-judging#480]
pinalikefruit

medium

# The owner/users must pay more than baseFee when creating a pool

When the `owner` tries to create a new pool, they always have to pay more than the `baseFee` and the protocol will always have to return the additional amount paid for each pool created.

## Vulnerability Detail

After a `owner` has registered, they will want to create a pool by calling `Allo.createPool()` or `Allo.createPoolWithCustomStrategy()`, after passing some checks, both call the internal function `Allo._createPool()`.

The Allo protocol exposed the problem of knowing that when the `baseFee` is 0, a pool can be created without paying it, but when it is not zero, it does not work as expected.

If `baseFee > 0`, the following is checked:

```solidity
Allo.sol:473     if ((_token == NATIVE && (baseFee + _amount >= msg.value)) || (_token != NATIVE && baseFee >= msg.value))
```
Even when the user sent the correct `msg.value` for the `baseFee` in whathever case, fund with `native` token and you dont send any `amount` or if you fund with amount but not a `native` token always will  return a error `NOT_ENOUGH_FUNDS()`, and this is not correct when the owner is pay exacly the fee ask for the protocol. 

Even when the user submitted the correct `msg.value` (i.e. `msg.value = baseFee`)  for the `baseFee` in any case funding with `NATIVE` token and not submitting any `amount` or funding with an amount but not a `NATIVE` token will always return an error ` NOT_ENOUGH_FUNDS()`, and this is not correct when the owner pays exactly the fee requested by the protocol.

Here I explain the error with a brief implementation:

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.19;

import "forge-std/Test.sol";
import { Allo } from "../../../contracts/core/Allo.sol";
import { Registry } from "../../../contracts/core/Registry.sol";
import "../../../contracts/core/libraries/Transfer.sol";

error NOT_ENOUGH_FUNDS();


contract MockAllo is Transfer{
    
    // the value of the `baseFee` was use in the test folder for the Allo protocol.
    uint baseFee = 1e15;

    function fundPool(address token, uint256 amount, address to) public payable {

        if (baseFee > 0) {
            if ((token == NATIVE && (baseFee + amount >= msg.value)) || (token != NATIVE && baseFee >= msg.value)) {
                revert NOT_ENOUGH_FUNDS();
            }
            _transferAmount(NATIVE, to, baseFee);
        }

    }
}


contract AlloTest is Test, Transfer {

    MockAllo public mockFund;
    address public  treasury = payable(makeAddr("treasury"));
    address bob = makeAddr("bob");
    address erc20 = makeAddr("erc20");

    function setUp() public {
        mockFund = new MockAllo();
    }

    function test_attackbase() public {
        vm.deal(bob, 100 ether);
        vm.startPrank(bob);
        // Try create a pool with native token
        vm.expectRevert(NOT_ENOUGH_FUNDS.selector);
        mockFund.fundPool{value:1e15}(NATIVE,0,treasury);
        
        // try crete a pool with erc-20 token
        vm.expectRevert(NOT_ENOUGH_FUNDS.selector);
        mockFund.fundPool{value:1e15}(erc20,0,treasury);
        
    }
      
}
```

## Impact

Every time an owner tries to create a new pool, they can try multiple times by paying the correct value of `baseFee`, but it will always throw an error `NOT_ENOUGH_FUNDS()`, the protocol may lose users, who will spend gas or for having to pay more than the stated rate.

On the other hand, if owners send more than the base rate value, the protocol must refund this payment for each group created. Because there is no function that returns the additional sent amount or a check that only the base fare amount is sent.

In short, the protocol could lose users and money by not accepting payment of the correct `baseFee` and by attempting to refund the amount sent for each pool created.

## Code Snippet
https://github.com/sherlock-audit/2023-09-Gitcoin/blob/main/allo-v2/contracts/core/Allo.sol#L473

## Tool used

* Manual Review
* Foundry

## Recommendation
Update the declararion for:
```diff
-   if ((token == NATIVE && (baseFee + amount >= msg.value)) || (token != NATIVE && baseFee >= msg.value));
+   if ((token == NATIVE && (baseFee + amount > msg.value)) || (token != NATIVE && baseFee > msg.value))
```
